solidity/contracts/AltheaDB.sol

```diff
@@ -1,54 +1,369 @@
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+        if (
+            !identities_are_equal(
+                get_registered_exit_with_wg_key(entry.wg_key),
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+
+        if (
+            !identities_are_equal(
+                get_registered_client_with_eth_addr(entry.eth_addr),
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+        if (
+            !identities_are_equal(
+                get_registered_client_with_mesh_ip(entry.mesh_ip),
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+        if (
+            !identities_are_equal(
+                get_registered_client_with_wg_key(entry.wg_key),
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    // start user and exit management functions
 
     // Add a new registered user
-    function add_registered_user(Identity memory entry) public {
-        registered_users.push(entry);
-        wg_key_to_reg_users_map[entry.wg_key] = entry;
-        mesh_ip_to_reg_users_map[entry.mesh_ip] = entry;
-        eth_addr_to_reg_users_map[entry.eth_addr] = entry;
+    function add_registered_user(Identity calldata entry) public {
+        if (is_user_admin(msg.sender)) {
+            // if any client or exit currently registered has overlapping data, do not allow the
+            // registration to continue
+            if (check_for_any_duplicates(entry)) {
+                revert DuplicateUser();
+            }
+
+            state_registeredUsers.push(entry);
+            emit UserRegisteredEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
+    }
+
+    // Remove a new registered user
+    function remove_registered_user(Identity calldata entry) public {
+        if (is_user_admin(msg.sender)) {
+            uint256 index = get_index_of_id(entry, state_registeredUsers);
+            delete_array_entry(index, state_registeredUsers);
+            emit UserRemovedEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
+    }
+
+    // Add a new registered exit
+    function add_registered_exit(Identity calldata entry) public {
+        if (is_exit_admin(msg.sender)) {
+            // if any client or exit currently registered has overlapping data, do not allow the
+            // registration to continue
+            if (check_for_any_duplicates(entry)) {
+                revert DuplicateUser();
+            }
+
+            state_registeredExits.push(entry);
+            emit ExitRegisteredEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
     }
 
-    // Get all registered users
-    function get_all_registered_users()
```

Comment:
This looks like this was removed and I dont see a replacement. Core requirement of rita exit is to get all users every tick, how will this be done?

Reply:
Note that I changed the users lists from private to public, public variables have automatic getters meaning they are simply accessible by calling the variable name as a function. Give it a try.

In this case 'public' means 'readable by other code' but does not allow anyone else to modify the data.